### ISSUE TITLE
`RecoverMultiplicationTable(8, 600*1000*1000);` fails

### ISSUE THREAD

**Issue by sweeneyde** [gap-packages/smallsemi#34]
I'm on Windows Subsystem for Linux, and this is what I'm seeing whenever I try to access certain multiplication tables:

<details>

```
dennis@DESKTOP-HSC9NKL:~/GAP$ ./gap-4.12.2_again/gap -m 1g
 ┌───────┐   GAP 4.12.2 of 2022-12-18
 │  GAP  │   https://www.gap-system.org
 └───────┘   Architecture: x86_64-pc-linux-gnu-default64-kv8
 Configuration:  gmp 6.2.1, GASMAN, readline
 Loading the library and packages ...
 Packages:   AClib 1.3.2, Alnuth 3.2.1, AtlasRep 2.1.6, AutPGrp 1.11, CRISP 1.4.6, Cryst 4.1.25, CrystCat 1.1.10, CTblLib 1.3.4, FactInt 1.6.3, FGA 1.4.0, GAPDoc 1.6.6, IRREDSOL 1.4.4, LAGUNA 3.9.5,
             Polenta 1.3.10, Polycyclic 2.16, PrimGrp 3.4.3, RadiRoot 2.9, ResClasses 4.7.3, SmallGrp 1.5.1, Sophus 1.27, SpinSym 1.5.2, TomLib 1.2.9, TransGrp 3.6.3, utils 0.81
 Try '??help' for help. See also '?copyright', '?cite' and '?authors'
gap> LoadPackage("smallsemi");
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Smallsemi -   A library of small semigroups
by Andreas Distler & James Mitchell
For contents, type: ?Smallsemi:
Loading Smallsemi 0.6.13 ...
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
true
gap> RecoverMultiplicationTable(8, 600*1000*1000);
#I  Smallsemi: loading data for semigroup properties. Please be patient.
#I  Smallsemi: loading data for semigroups of size 8.
Error, in StringFile: no error (0)
 at /home/dennis/GAP/gap-4.12.2_again/lib/string.gi:748 called from
StringFile( file ) at /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/gap/small.gi:686 called from
READ_3NIL_DATA( diag ); at /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/gap/small.gi:423 called from
RecoverMultiplicationTableNC( size, nr ) at /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/gap/small.gi:337 called from
<function "RecoverMultiplicationTable">( <arguments> )
 called from read-eval loop at *stdin*:2
you can 'quit;' to quit to outer loop, or
you can 'return;' to continue
brk> file
> ;
"/home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/data/data8-3nil/diag11223.gl"
```

</details>

It looks like `StringFile("/home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/data/data8-3nil/diag11223.gl")` fails similarly on its own as well. It looks like I have some `gl.gz` files available, but none of the unpacked `.gl` files:

<details>

```
dennis@DESKTOP-HSC9NKL:~/GAP$ ls /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/data/data8-3nil/
diag11.gl.gz         diag11112.gl.gz      diag11122pos.gl.gz  diag112222pos.gl.gz  diag12.gl.gz         diag1222pos.gl.gz   diag123pos.gl.gz     diag22222pos.gl.gz  diag223.gl.gz      diag233pos.gl.gz
diag111.gl.gz        diag111122.gl.gz     diag11123.gl.gz     diag11222pos.gl.gz   diag122.gl.gz        diag1223.gl.gz      diag12pos.gl.gz      diag22223.gl.gz     diag2233.gl.gz     diag234.gl.gz
diag1111.gl.gz       diag111122pos.gl.gz  diag11123pos.gl.gz  diag11223.gl.gz      diag1222.gl.gz       diag12233.gl.gz     diag2.gl.gz          diag22223pos.gl.gz  diag2233pos.gl.gz  diag234pos.gl.gz
diag11111.gl.gz      diag11112pos.gl.gz   diag1112pos.gl.gz   diag11223pos.gl.gz   diag12222.gl.gz      diag12233pos.gl.gz  diag22.gl.gz         diag2222pos.gl.gz   diag2234.gl.gz     diag23pos.gl.gz
diag111111.gl.gz     diag1111pos.gl.gz    diag111pos.gl.gz    diag1122pos.gl.gz    diag122222.gl.gz     diag1223pos.gl.gz   diag222.gl.gz        diag2223.gl.gz      diag2234pos.gl.gz  diag2pos.gl.gz
diag111111pos.gl.gz  diag1112.gl.gz       diag112.gl.gz       diag1123.gl.gz       diag122222pos.gl.gz  diag122pos.gl.gz    diag2222.gl.gz       diag22233.gl.gz     diag223pos.gl.gz
diag111112.gl.gz     diag11122.gl.gz      diag1122.gl.gz      diag1123pos.gl.gz    diag12222pos.gl.gz   diag123.gl.gz       diag22222.gl.gz      diag22233pos.gl.gz  diag22pos.gl.gz
diag111112pos.gl.gz  diag111222.gl.gz     diag11222.gl.gz     diag112pos.gl.gz     diag12223.gl.gz      diag1234.gl.gz      diag222222.gl.gz     diag2223pos.gl.gz   diag23.gl.gz
diag11111pos.gl.gz   diag111222pos.gl.gz  diag112222.gl.gz    diag11pos.gl.gz      diag12223pos.gl.gz   diag1234pos.gl.gz   diag222222pos.gl.gz  diag222pos.gl.gz    diag233.gl.gz
```
</details>

Is there some sort of issue with caching the unpacked files? Am I doing something wrong?

All of the tests pass:
<details>

```
gap> ReadPackage( "smallsemi", "tst/testall.g" );
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Smallsemi -   A library of small semigroups
by Andreas Distler & James Mitchell
For contents, type: ?Smallsemi:
Loading Smallsemi 0.6.13 ...
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Architecture: x86_64-pc-linux-gnu-default64-kv8

testing: /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/tst/coclass.tst
    3335 ms (134 ms GC) and 472MB allocated for coclass.tst
testing: /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/tst/enums.tst
    6783 ms (424 ms GC) and 2.69GB allocated for enums.tst
testing: /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/tst/greensstar.tst
     450 ms (308 ms GC) and 90.2MB allocated for greensstar.tst
testing: /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/tst/properties.tst
     785 ms (122 ms GC) and 213MB allocated for properties.tst
testing: /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/tst/small.tst
     435 ms (138 ms GC) and 174MB allocated for small.tst
testing: /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/tst/smallsemi01.tst
     659 ms (139 ms GC) and 128MB allocated for smallsemi01.tst
testing: /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/tst/smallsemi02.tst
    3169 ms (155 ms GC) and 1.21GB allocated for smallsemi02.tst
-----------------------------------
total     15616 ms (1420 ms GC) and 4.96GB allocated
              0 failures in 7 files

#I  No errors detected while testing
```
</details>

Thanks for your help

**Comment by sweeneyde** [comment]
It seems to depend on the access order within a GAP session:

* Accessing 600Mil-th 8x8 multiplication table fails
* Then accessing 500Mil-th succeeds as usual
* Then accessing 1Bil-th fails
* Then accessing 1Bil-th succeeds
* Then accessing 600Mil-th finally succeeds

<details>

```
gap> RecoverMultiplicationTable(8, 600*1000*1000);
#I  Smallsemi: loading data for semigroup properties. Please be patient.
#I  Smallsemi: loading data for semigroups of size 8.
Error, in StringFile: no error (0)
 at /home/dennis/GAP/gap-4.12.2_again/lib/string.gi:748 called from
StringFile( file ) at /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/gap/small.gi:686 called from
READ_3NIL_DATA( diag ); at /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/gap/small.gi:423 called from
RecoverMultiplicationTableNC( size, nr ) at /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/gap/small.gi:337 called from
<function "RecoverMultiplicationTable">( <arguments> )
 called from read-eval loop at *stdin*:2
you can 'quit;' to quit to outer loop, or
you can 'return;' to continue
brk> quit;
gap> RecoverMultiplicationTable(8, 500*1000*1000);
#I  Smallsemi: loading data for semigroups of size 8.
[ [ 1, 1, 1, 1, 1, 1, 1, 1 ], [ 1, 1, 1, 1, 1, 1, 1, 1 ], [ 1, 1, 1, 1, 1, 1, 1, 1 ], [ 1, 1, 1, 1, 1, 1, 2, 3 ],
  [ 1, 1, 1, 3, 1, 3, 2, 3 ], [ 1, 1, 1, 3, 3, 2, 1, 1 ], [ 1, 1, 1, 3, 1, 2, 2, 1 ], [ 1, 1, 1, 2, 3, 1, 3, 2 ] ]
gap> RecoverMultiplicationTable(8, 1000*1000*1000);
#I  Smallsemi: loading data for semigroups of size 8.
Error, in StringFile: no error (0)
 at /home/dennis/GAP/gap-4.12.2_again/lib/string.gi:748 called from
StringFile( file ) at /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/gap/small.gi:686 called from
READ_3NIL_DATA( diag ); at /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/gap/small.gi:423 called from
RecoverMultiplicationTableNC( size, nr ) at /home/dennis/GAP/gap-4.12.2_again/pkg/smallsemi/gap/small.gi:337 called from
<function "RecoverMultiplicationTable">( <arguments> )
 called from read-eval loop at *stdin*:3
you can 'quit;' to quit to outer loop, or
you can 'return;' to continue
brk> quit;
gap> RecoverMultiplicationTable(8, 1000*1000*1000);
[ [ 1, 1, 1, 1, 1, 1, 1, 1 ], [ 1, 1, 1, 1, 1, 1, 1, 1 ], [ 1, 1, 1, 1, 1, 1, 1, 1 ], [ 1, 1, 1, 1, 1, 1, 3, 1 ],
  [ 1, 1, 1, 2, 1, 2, 1, 1 ], [ 1, 1, 1, 2, 3, 2, 1, 1 ], [ 1, 1, 1, 3, 3, 2, 2, 3 ], [ 1, 1, 1, 1, 3, 3, 3, 3 ] ]
gap> RecoverMultiplicationTable(8, 600*1000*1000);
[ [ 1, 1, 1, 1, 1, 1, 1, 1 ], [ 1, 1, 1, 1, 1, 1, 1, 1 ], [ 1, 1, 1, 1, 1, 1, 1, 1 ], [ 1, 1, 1, 1, 1, 1, 1, 2 ],
  [ 1, 1, 1, 3, 1, 2, 3, 1 ], [ 1, 1, 1, 1, 1, 2, 3, 2 ], [ 1, 1, 1, 1, 1, 3, 2, 2 ], [ 1, 1, 1, 2, 3, 1, 1, 3 ] ]
```
</details>

That exact procedure actually seems to be repeatable for me.

The lowest-index call I could find that fails is this: `RecoverMultiplicationTable(8, 576676952);`. One less (576676951) seems to always succeed.

**Comment by james-d-mitchell** [comment]
Thanks for the report @sweeneyde, I can reproduce this, and will investigate. I don't think you're doing anything wrong, seems that something isn't working properly.

**Comment by james-d-mitchell** [comment]
Just to check can you please indicate which version of GAP you are using and which version of Smallsemi? Thanks!

**Comment by james-d-mitchell** [comment]
Ok, this bug didn't exist in GAP 4.10.2 but did exist in GAP 4.11.1, I think this is a bug in GAP not Smallsemi but will investigate some more and try to pin down exactly what's going on.

**Comment by james-d-mitchell** [comment]
Ah, no, a quick bisect indicates that the commit that introduced this bug was:

aa83955823b3c453bcedec551c2417e5a7e00538

it seems that when removing the remnants of CVS, two of the data files were also corrupted. This is an easy fix!

**Comment by james-d-mitchell** [comment]
For the record this commit was made between versions 0.6.11 and 0.6.13 of Smallsemi. Version 0.6.11 was included with GAP 4.10.2, and version 0.6.13 with GAP 4.11.1, which explains my comments above.

**Comment by james-d-mitchell** [comment]
I'm guessing what has happened here is that when removing the remnants of CVS from the repo @fingolfin, did some kind of search and replace for (something related to) `## $Id$`, and that this pattern was coincidentally in two of the 'gz' data files for Smallsemi. The strings in the data files where replaced, rendering the gzip files corrupt, which is why some tables were inaccessible. Just a guess!
